src/Internal/Contract/QueryBackend.purs

```diff
@@ -0,0 +1,156 @@
+module Ctl.Internal.Contract.QueryBackend
+  ( BlockfrostBackend
+  , CtlBackend
+  , QueryBackend(BlockfrostBackend, CtlBackend)
+  , QueryBackendLabel(BlockfrostBackendLabel, CtlBackendLabel)
+  , QueryBackendParams(BlockfrostBackendParams, CtlBackendParams)
+  , QueryBackends
+  , class HasQueryBackendLabel
+  , backendLabel
+  , defaultBackend
+  , lookupBackend
+  , mkBackendParams
+  , mkSingletonBackendParams
+  , mkCtlBackendParams
+  , mkBlockfrostBackendParams
+  ) where
+
+import Prelude
+
+import Ctl.Internal.QueryM (DatumCacheWebSocket, OgmiosWebSocket)
+import Ctl.Internal.QueryM.ServerConfig (ServerConfig)
+import Data.Array (nub) as Array
+import Data.Array ((:))
+import Data.Foldable (class Foldable, foldMap, foldl, foldr, length)
+import Data.Map (Map)
+import Data.Map (empty, insert, lookup) as Map
+import Data.Maybe (Maybe(Just))
+import Data.Traversable (class Traversable, sequence, traverse)
+import Effect (Effect)
+import Effect.Exception (throw)
+
+--------------------------------------------------------------------------------
+-- QueryBackends
+--------------------------------------------------------------------------------
+
+-- | A generic type to represent a choice of backend with a set of fallback
+-- | backends when an operation is not supported by the default.
+data QueryBackends (backend :: Type) =
+  QueryBackends backend (Map QueryBackendLabel backend)
+
+-- Functor breaks this datatype...
+derive instance Functor QueryBackends
+
+instance Foldable QueryBackends where
+  foldr f z (QueryBackends x xs) = f x (foldr f z xs)
+  foldl f z (QueryBackends x xs) = foldl f (f z x) xs
+  foldMap f (QueryBackends x xs) = f x <> foldMap f xs
+
+instance Traversable QueryBackends where
+  traverse f (QueryBackends x xs) = QueryBackends <$> f x <*> traverse f xs
+  sequence (QueryBackends x xs) = QueryBackends <$> x <*> sequence xs
+
+mkSingletonBackendParams
+  :: QueryBackendParams -> QueryBackends QueryBackendParams
+mkSingletonBackendParams = flip QueryBackends Map.empty
+
+mkCtlBackendParams
+  :: { ogmiosConfig :: ServerConfig
+     , kupoConfig :: ServerConfig
+     , odcConfig :: ServerConfig
+     }
+  -> QueryBackends QueryBackendParams
+mkCtlBackendParams = mkSingletonBackendParams <<< CtlBackendParams
+
+mkBlockfrostBackendParams :: ServerConfig -> QueryBackends QueryBackendParams
+mkBlockfrostBackendParams = mkSingletonBackendParams <<< BlockfrostBackendParams
+  <<< { blockfrostConfig: _ }
+
+mkBackendParams
+  :: QueryBackendParams
+  -> Array QueryBackendParams
+  -> Effect (QueryBackends QueryBackendParams)
+mkBackendParams defaultBackend' backends =
```

Comment:
These `mk*BackendParams` functions return `QueryBackends`, might be clearer to rename them to `mk*QueryBackends`. Or `mk*Backend` might make sense because we refer to `QueryBackends QueryBackend` as `backend` [here](https://github.com/Plutonomicon/cardano-transaction-lib//blob/5857e5eb826b27ba484c6ef36ee5ff3b8484a072/src/Internal/Contract/Monad.purs#L166 "ContractEnv type")